sw/device/tests/sim_dv/sysrst_ctrl_ulp_z3_wakeup_test.c

```diff
@@ -0,0 +1,192 @@
+#include "sw/device/lib/dif/dif_sysrst_ctrl.h"
+#include "sw/device/lib/runtime/ibex.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/pwrmgr_testutils.h"
+#include "sw/device/lib/testing/rstmgr_testutils.h"
+#include "sw/device/lib/testing/test_framework/check.h"
+#include "sw/device/lib/testing/test_framework/ottf_main.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+OTTF_DEFINE_TEST_CONFIG();
+
+// This is updated by the sv component of the test
+// Now `kTestPhase` is relocated to RamMain0
+static volatile uint8_t kTestPhase = 0;
+
+static dif_pwrmgr_t pwrmgr;
+static dif_rstmgr_t rstmgr;
+static dif_pinmux_t pinmux;
+static dif_sysrst_ctrl_t sysrst_ctrl;
+
+// This means 20 aon_clk ticks ~= 20 * 5 us = 100 us
+static const uint16_t debounce_timer = 20;
+
+enum {
+  kNumMioInPads = 3,
+  kNumMioOutPads = 1,
+};
+
+const uint32_t kTestPhaseTimeoutUsec = 500;
+enum {
+  kTestPhaseInit = 0,
+  kTestPhaseDriveZero = 1,
+  kTestPhaseWaitNoWakeup = 2,
+  kTestPhaseGlitchLidOpen = 3,
+  kTestPhaseWaitWakeup = 4,
+  kTestPhaseDone = 5,
+};
+
+static const dif_pinmux_index_t kPeripheralInputs[] = {
+    kTopEarlgreyPinmuxPeripheralInSysrstCtrlAonPwrbIn,
+    kTopEarlgreyPinmuxPeripheralInSysrstCtrlAonAcPresent,
+    kTopEarlgreyPinmuxPeripheralInSysrstCtrlAonLidOpen,
+};
+
+static const dif_pinmux_index_t kInputPads[] = {
+    kTopEarlgreyPinmuxInselIor13,
+    kTopEarlgreyPinmuxInselIoc7,
+    kTopEarlgreyPinmuxInselIoc9,
+};
+
+static const dif_pinmux_index_t kPeripheralOutputs[] = {
+    kTopEarlgreyPinmuxOutselSysrstCtrlAonZ3Wakeup,
+};
+
+static const dif_pinmux_index_t kOutputPads[] = {
+    kTopEarlgreyPinmuxMioOutIob7,
+};
+
+/**
+ * Sets up the pinmux to assign input and output pads to the sysrst_ctrl
+ * peripheral as required.
+ */
+static void pinmux_setup(void) {
+  for (int i = 0; i < kNumMioInPads; ++i) {
+    CHECK_DIF_OK(
+        dif_pinmux_input_select(&pinmux, kPeripheralInputs[i], kInputPads[i]));
+  }
+
+  for (int i = 0; i < kNumMioOutPads; ++i) {
+    CHECK_DIF_OK(dif_pinmux_output_select(&pinmux, kOutputPads[i],
+                                          kPeripheralOutputs[i]));
+  }
+}
+
+/**
+ * Waits for `kTestPhase` variable to be changed by a backdoor overwrite
+ * from the testbench in chip_sw_<testname>_vseq.sv. This will indicate that
+ * the testbench is ready to proceed with the next phase of the test.
+ */
+static void wait_next_test_phase(void) {
+  uint8_t current_phase = kTestPhase;
+  // Set WFI status for testbench synchronization
+  // No WFI instruction is issued
+  LOG_INFO("Prev test phase = %0d", kTestPhase);
+  test_status_set(kTestStatusInWfi);
+  test_status_set(kTestStatusInTest);
+  // IBEX_SPIN_FOR(current_phase != kTestPhase, kTestPhaseTimeoutUsec);
+  const ibex_timeout_t timeout_ = ibex_timeout_init(kTestPhaseTimeoutUsec);
+  while (current_phase == kTestPhase) {
+    CHECK(!ibex_timeout_check(&timeout_), "Timed out\n");
+    //kTestPhase = 0xff;
+  }
+
```

Comment:
This is where I unrolled IBEX_SPIN_FOR. When I run this test with dvsim, the loop terminates with timeout error. I didn't overwrite `kTestPhase` with `0xff`, because this variable lives in RamMain0.